<a href="https://colab.research.google.com/github/Stark-Logan/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/LS_DS_214_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 4*

---

# Logistic Regression


## Assignment 🌯

You'll use a [**dataset of 400+ burrito reviews**](https://srcole.github.io/100burritos/). How accurately can you predict whether a burrito is rated 'Great'?

> We have developed a 10-dimensional system for rating the burritos in San Diego. ... Generate models for what makes a burrito great and investigate correlations in its dimensions.

- [ ] Do train/validate/test split. Train on reviews from 2016 & earlier. Validate on 2017. Test on 2018 & later.
- [ ] Begin with baselines for classification.
- [ ] Use scikit-learn for logistic regression.
- [ ] Get your model's validation accuracy. (Multiple times if you try multiple iterations.)
- [ ] Get your model's test accuracy. (One time, at the end.)
- [ ] Commit your notebook to your fork of the GitHub repo.


## Stretch Goals

- [ ] Add your own stretch goal(s) !
- [ ] Make exploratory visualizations.
- [ ] Do one-hot encoding.
- [ ] Do [feature scaling](https://scikit-learn.org/stable/modules/preprocessing.html).
- [ ] Get and plot your coefficients.
- [ ] Try [scikit-learn pipelines](https://scikit-learn.org/stable/modules/compose.html).

In [1]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Linear-Models/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'

In [2]:
# Load data downloaded from https://srcole.github.io/100burritos/
import pandas as pd
df = pd.read_csv(DATA_PATH+'burritos/burritos.csv')

In [3]:
# Derive binary classification target:
# We define a 'Great' burrito as having an
# overall rating of 4 or higher, on a 5 point scale.
# Drop unrated burritos.
df = df.dropna(subset=['overall'])
df['Great'] = df['overall'] >= 4

In [4]:
# Clean/combine the Burrito categories
df['Burrito'] = df['Burrito'].str.lower()

california = df['Burrito'].str.contains('california')
asada = df['Burrito'].str.contains('asada')
surf = df['Burrito'].str.contains('surf')
carnitas = df['Burrito'].str.contains('carnitas')

df.loc[california, 'Burrito'] = 'California'
df.loc[asada, 'Burrito'] = 'Asada'
df.loc[surf, 'Burrito'] = 'Surf & Turf'
df.loc[carnitas, 'Burrito'] = 'Carnitas'
df.loc[~california & ~asada & ~surf & ~carnitas, 'Burrito'] = 'Other'

In [5]:
# Drop some high cardinality categoricals
df = df.drop(columns=['Notes', 'Location', 'Reviewer', 'Address', 'URL', 'Neighborhood'])

In [6]:
# Drop some columns to prevent "leakage"
df = df.drop(columns=['Rec', 'overall'])

In [7]:
df

Burrito       Date  Yelp  Google  ... Avocado  Corn  Zucchini  Great
0    California  1/18/2016   3.5     4.2  ...     NaN   NaN       NaN  False
1    California  1/24/2016   3.5     3.3  ...     NaN   NaN       NaN  False
2      Carnitas  1/24/2016   NaN     NaN  ...     NaN   NaN       NaN  False
3         Asada  1/24/2016   NaN     NaN  ...     NaN   NaN       NaN  False
4    California  1/27/2016   4.0     3.8  ...     NaN   NaN       NaN   True
..          ...        ...   ...     ...  ...     ...   ...       ...    ...
418       Other  8/27/2019   NaN     NaN  ...     NaN   NaN       NaN  False
419       Other  8/27/2019   NaN     NaN  ...     NaN   NaN       NaN   True
420  California  8/27/2019   NaN     NaN  ...     NaN   NaN       NaN  False
421       Other  8/27/2019   NaN     NaN  ...     NaN   NaN       NaN   True
422       Other  8/27/2019   NaN     NaN  ...     NaN   NaN       NaN   True

[421 rows x 59 columns]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 421 entries, 0 to 422
Data columns (total 59 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Burrito         421 non-null    object 
 1   Date            421 non-null    object 
 2   Yelp            87 non-null     float64
 3   Google          87 non-null     float64
 4   Chips           26 non-null     object 
 5   Cost            414 non-null    float64
 6   Hunger          418 non-null    float64
 7   Mass (g)        22 non-null     float64
 8   Density (g/mL)  22 non-null     float64
 9   Length          283 non-null    float64
 10  Circum          281 non-null    float64
 11  Volume          281 non-null    float64
 12  Tortilla        421 non-null    float64
 13  Temp            401 non-null    float64
 14  Meat            407 non-null    float64
 15  Fillings        418 non-null    float64
 16  Meat:filling    412 non-null    float64
 17  Uniformity      419 non-null    flo

In [9]:
df1 = df.dropna(thresh=370, axis=1)

In [10]:
df1

Burrito       Date  Cost  Hunger  ...  Salsa  Synergy  Wrap  Great
0    California  1/18/2016  6.49     3.0  ...    4.0      4.0   4.0  False
1    California  1/24/2016  5.45     3.5  ...    3.5      2.5   5.0  False
2      Carnitas  1/24/2016  4.85     1.5  ...    3.0      3.0   5.0  False
3         Asada  1/24/2016  5.25     2.0  ...    4.0      4.0   5.0  False
4    California  1/27/2016  6.59     4.0  ...    2.5      4.5   4.0   True
..          ...        ...   ...     ...  ...    ...      ...   ...    ...
418       Other  8/27/2019  6.00     1.0  ...    2.0      2.0   5.0  False
419       Other  8/27/2019  6.00     4.0  ...    5.0      4.0   3.0   True
420  California  8/27/2019  7.90     3.0  ...    3.5      4.0   4.5  False
421       Other  8/27/2019  7.90     3.0  ...    5.0      5.0   2.0   True
422       Other  8/27/2019  5.50     3.5  ...    3.0      4.5   4.0   True

[421 rows x 14 columns]

In [11]:
df1['Date'] = pd.to_datetime(df1['Date'], infer_datetime_format= True)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


#  Do train/validate/test split. Train on reviews from 2016 & earlier. Validate on 2017. Test on 2018 & later.

In [12]:
X_train = df1[df1['Date'] < '2017']

In [35]:
X_train = X_train.drop('Date', axis = 1)

In [47]:
X_train

Burrito  Cost  Hunger  Tortilla  ...  Uniformity  Salsa  Synergy  Wrap
0    California  6.49     3.0       3.0  ...         4.0    4.0      4.0   4.0
1    California  5.45     3.5       2.0  ...         4.0    3.5      2.5   5.0
2      Carnitas  4.85     1.5       3.0  ...         4.0    3.0      3.0   5.0
3         Asada  5.25     2.0       3.0  ...         5.0    4.0      4.0   5.0
4    California  6.59     4.0       4.0  ...         5.0    2.5      4.5   4.0
..          ...   ...     ...       ...  ...         ...    ...      ...   ...
296  California  5.65     3.0       4.0  ...         4.0    3.0      2.0   4.5
297       Other  5.49     3.0       4.5  ...         3.5    3.0      2.5   3.0
298  California  7.75     4.0       3.5  ...         2.3    2.2      3.3   4.5
299       Asada  7.75     4.0       4.0  ...         3.5    2.0      2.0   4.0
300       Other  6.99     3.7       3.6  ...         4.3    NaN      3.8   2.0

[298 rows x 12 columns]

In [14]:
X_valid = df1[(df1['Date'] < '2018') & (df1['Date'] >= '2017')] 

In [36]:
X_valid = X_valid.drop('Date', axis = 1)

In [48]:
X_valid

Burrito   Cost  Hunger  Tortilla  ...  Uniformity  Salsa  Synergy  Wrap
301  California   6.60     NaN       4.0  ...         5.0    1.5     3.50   4.5
302       Other   6.60     NaN       4.0  ...         4.6    4.2     3.75   5.0
303       Other   8.50     3.9       3.0  ...         4.0    4.3     4.20   5.0
304       Other   7.90     4.0       3.5  ...         4.5    4.0     3.80   4.8
305       Other   4.99     3.5       2.5  ...         3.0    2.0     2.00   4.0
..          ...    ...     ...       ...  ...         ...    ...      ...   ...
381  California   6.85     3.5       3.0  ...         3.5    3.5     4.00   3.0
382  California   6.85     3.5       3.0  ...         3.0    4.0     4.00   3.5
383       Other  11.50     3.5       2.0  ...         4.5    3.5     4.00   2.0
384  California   7.89     4.0       4.0  ...         4.0    3.5     4.30   4.5
385  California   7.89     5.0       5.0  ...         4.0    5.0     5.00   3.0

[85 rows x 12 columns]

In [16]:
X_test = df1[df1['Date'] >= '2018']

In [34]:
X_test = X_test.drop('Date', axis = 1)

In [50]:
X_test.shape

(38, 12)

In [18]:
df1[df1['Date'] >= '2026']

Burrito       Date  Cost  Hunger  ...  Salsa  Synergy  Wrap  Great
77  California 2026-04-25   8.0     4.0  ...    3.0      5.0   5.0   True

[1 rows x 14 columns]

# Begin with baselines for classification.

In [19]:
y_train = X_train.pop('Great')

In [20]:
y_valid = X_valid.pop('Great')

In [21]:
y_test = X_test.pop('Great')

In [22]:
y_train.value_counts(normalize=True)

False    0.590604
True     0.409396
Name: Great, dtype: float64

-Our baseline is .590604

 # Use scikit-learn for logistic regression.

In [23]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from category_encoders import OneHotEncoder 
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [24]:
df1.describe(exclude= 'number')

Burrito                 Date  Great
count          421                  421    421
unique           5                  169      2
top     California  2016-08-30 00:00:00  False
freq           169                   29    239
first          NaN  2011-05-16 00:00:00    NaN
last           NaN  2026-04-25 00:00:00    NaN

In [37]:
logistic_model = Pipeline([
                   ('ohe' , OneHotEncoder(cols = ['Burrito'])),
                   ('imputer' , SimpleImputer()),
                   ('classifier' , LogisticRegression())

])

In [44]:
logistic_model.fit(X_train,y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


Pipeline(memory=None,
         steps=[('ohe',
                 OneHotEncoder(cols=['Burrito'], drop_invariant=False,
                               handle_missing='value', handle_unknown='value',
                               return_df=True, use_cat_names=False,
                               verbose=0)),
                ('imputer',
                 SimpleImputer(add_indicator=False, copy=True, fill_value=None,
                               missing_values=nan, strategy='mean',
                               verbose=0)),
                ('classifier',
                 LogisticRegression(C=1.0, class_weight=None, dual=False,
                                    fit_intercept=True, intercept_scaling=1,
                                    l1_ratio=None, max_iter=100,
                                    multi_class='auto', n_jobs=None,
                                    penalty='l2', random_state=None,
                                    solver='lbfgs', tol=0.0001, verbose=0,
            

In [45]:
print('Training score:' ,logistic_model.score(X_train,y_train))
print('Validation score:', logistic_model.score(X_valid, y_valid))

Training score: 0.889261744966443
Validation score: 0.8352941176470589


In [46]:
X_train.shape

(298, 12)

In [52]:
print('Testing score:', logistic_model.score(X_test, y_test))

Testing score: 0.7631578947368421
